src/mlpack/methods/lmnn/constraints.hpp

```diff
@@ -0,0 +1,188 @@
+/**
+ * @file constraints.hpp
+ * @author Manish Kumar
+ *
+ * Declaration of the Constraints class.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_LMNN_CONSTRAINTS_HPP
+#define MLPACK_METHODS_LMNN_CONSTRAINTS_HPP
+
+#include <mlpack/prereqs.hpp>
+#include <mlpack/methods/neighbor_search/neighbor_search.hpp>
+
+namespace mlpack {
+namespace lmnn {
+/**
+ * Interface for generating distance based constraints on a given
+ * dataset, provided corresponding true labels and a quantity parameter (k)
+ * are specified.
+ *
+ * Class provides TargetNeighbors() (Used for calculating target neighbors
+ * of each data point), Impostors() (used for calculating impostors of each
+ * data point) and Triplets() (Generates sets of {dataset, target neighbors,
+ * impostors} tripltets.)
+ */
+template<typename MetricType = metric::SquaredEuclideanDistance>
+class Constraints
+{
+ public:
+  //! Convenience typedef.
+  typedef neighbor::NeighborSearch<neighbor::NearestNeighborSort, MetricType>
+      KNN;
+
+  /**
+   * Constructor for creating a Constraints instance.
+   *
+   * @param dataset Input dataset.
+   * @param labels Input dataset labels.
+   * @param k Number of target neighbors, impostors & triplets.
+   */
+  Constraints(const arma::mat& dataset,
+              const arma::Row<size_t>& labels,
+              size_t k);
```

Comment:
Should we use ``const`` here as well, to be consistent with the rest of the codebase?